api/output_policy.go

```diff
@@ -0,0 +1,197 @@
+package api
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+
+	retryablehttp "github.com/hashicorp/go-retryablehttp"
+	"github.com/hashicorp/vault/sdk/helper/jsonutil"
+)
+
+const (
+	ErrOutputPolicyRequest = "output policy request"
+)
+
+var LastOutputPolicyError *OutputPolicyError
+
+type OutputPolicyError struct {
+	*retryablehttp.Request
+	VaultClient     *Client
+	parsingError    error
+	parsedHCLString string
+}
+
+func (d *OutputPolicyError) Error() string {
+	if d.parsedHCLString == "" {
+		d.parseRequest()
+		if d.parsingError != nil {
+			return d.parsingError.Error()
+		}
+	}
+
+	return ErrOutputPolicyRequest
+}
+
+// Builds a sample policy document from the request
+func (d *OutputPolicyError) parseRequest() {
+	capabilities := []string{}
+	switch d.Request.Method {
+	case "GET":
+		capabilities = append(capabilities, "read")
+	case "LIST":
+		capabilities = append(capabilities, "list")
+	case "POST", "PUT":
+		capabilities = append(capabilities, "create")
+		capabilities = append(capabilities, "update")
+	case "PATCH":
+		capabilities = append(capabilities, "patch")
+	case "DELETE":
+		capabilities = append(capabilities, "delete")
+	}
+
+	// trim the Vault address and v1 from the front of the path
+	url := d.Request.URL.String()
+	apiAddrPrefix := fmt.Sprintf("%sv1/", d.VaultClient.config.Address)
+	path := strings.Trim(url, apiAddrPrefix)
+
+	// determine whether to add sudo capability
+	needsSudo, err := isSudoPath(d.VaultClient, path)
+	if err != nil {
+		d.parsingError = err
+		return
+	}
+	if needsSudo {
+		capabilities = append(capabilities, "sudo")
+	}
+
+	capStr := strings.Join(capabilities, `", "`)
+	d.parsedHCLString = fmt.Sprintf(
+		`path "%s" {
+  capabilities = ["%s"]
+}`, path, capStr)
+}
+
+func (d *OutputPolicyError) HCLString() string {
+	if d.parsedHCLString == "" {
+		d.parseRequest()
+	}
+	return d.parsedHCLString
+}
+
+// Determine whether the given path requires the sudo capability
+func isSudoPath(client *Client, path string) (bool, error) {
+	sudoPaths, err := getSudoPaths(client)
```

Comment:
Do we know how many sudo paths there are? If I understand correctly, there will be (1 + sudoPath) API requests made per call here, which could be quite expensive. If this is a static list of things, we could possibly hard code them or source them from elsewhere.

Reply:
I had started with hard-coding but some discussion in the RFC made me reconsider it, as it would be hard to remember to keep it up to date as we add new sudo endpoints.

And can you clarify what you mean here by the expensive multiple API requests? As far as I can tell it's just the one API request, since we're only hitting the OpenAPI endpoint one time and then comparing against the response object after that.

Reply:
Sorry, I might have misread the code -- somehow I thought that we were making a new request for each sudo path.

> I had started with hard-coding but some discussion in the RFC made me reconsider it, as it would be hard to remember to keep it up to date as we add new sudo endpoints.

One option to consider is to introduce a unit test which would ensure that your static list of sudo paths matches exactly with all the sudo paths that are defined in OpenAPI spec. This would essentially shift the performance hit from run-time to test time. It would also avoid the pitfalls of regexes mentioned in the other comment.

Reply:
To further @averche's suggestion of a static list checked during testing, if you choose to go that way,  I believe it would also remove the need to store a client pointer in the error.